src/geoarrow.cpp

```diff
@@ -71,7 +71,178 @@ py::array_t<PyObjectGeography> from_geoarrow(py::object input,
+        schema_src->release = nullptr;
+    }
+
+    py::tuple return_capsules(py::args args, const py::kwargs& kwargs) {
+        if ((args.size() > 0) && (!args[0].is_none())) {
+            throw std::invalid_argument(
+                "Additional arguments (such as requested_schema) with a non-default value are not "
+                "supported");
+        }
+        if (kwargs) {
+            for (auto& item : kwargs) {
+                if (!item.second.is_none()) {
+                    throw std::invalid_argument(
+                        "Additional arguments (such as requested_schema) with a non-default value "
+                        "are not supported");
+                }
+            }
+        }
+
+        ArrowArray* c_array = static_cast<ArrowArray*>(malloc(sizeof(ArrowArray)));
+        ArrowSchema* c_schema = static_cast<ArrowSchema*>(malloc(sizeof(ArrowSchema)));
+        move(&array_, &schema_, c_array, c_schema);
+
+        constexpr auto array_cleanup = [](void* ptr) noexcept {
+            auto array = static_cast<ArrowArray*>(ptr);
+            if (array->release != nullptr) {
+                array->release(array);
+            }
+            free(array);
+        };
+        py::capsule array_capsule{c_array, "arrow_array", array_cleanup};
+
+        constexpr auto schema_cleanup = [](void* ptr) noexcept {
+            auto schema = static_cast<ArrowSchema*>(ptr);
+            if (schema->release != nullptr) {
+                schema->release(schema);
+            }
+            free(schema);
+        };
+        py::capsule schema_capsule{c_schema, "arrow_schema", schema_cleanup};
+
+        return py::make_tuple(schema_capsule, array_capsule);
+    }
+
+    void move(ArrowArray* array_src,
+              ArrowSchema* schema_src,
+              ArrowArray* array_dst,
+              ArrowSchema* schema_dst) {
+        memcpy(array_dst, array_src, sizeof(struct ArrowArray));
+        array_src->release = nullptr;
+
+        memcpy(schema_dst, schema_src, sizeof(struct ArrowSchema));
+        schema_src->release = nullptr;
+    }
+
+    /// \brief Call data's release callback if valid
+    void reset() {
+        if (array_.release != nullptr) {
+            array_.release(&array_);
+        }
+
+        if (schema_.release != nullptr) {
+            schema_.release(&schema_);
+        }
+    }
+
+    /// \brief Call data's release callback if valid and move ownership of the data
+    /// pointed to by data
+    void reset(ArrowArray* array_src, ArrowSchema* schema_src) {
+        reset();
+        move(array_src, schema_src, &array_, &schema_);
+    }
+
+    ~ArrowArrayHolder() {
+        reset();
+    }
+
+protected:
+    ArrowArray array_;
+    ArrowSchema schema_;
+};
+
+ArrowArrayHolder to_geoarrow(py::array_t<PyObjectGeography> input,
+                             py::object output_schema,
+                             //  py::object geometry_encoding,
```

Comment:
Is it left for a follow-up PR?

Reply:
Yeah, just as for the input function, for WKB/WKT I would like that the user could just specify that without having the create a matching arrow schema, for convenience. But s2geography does not yet expose the necessary tools to do so (it already allow to specify WKT/WKB as output type, but then we would still have to create a ArrowSchema struct to create a valid output `ArrowArrayHolder`)

Reply:
But I can also leave this commented part out to clean things up